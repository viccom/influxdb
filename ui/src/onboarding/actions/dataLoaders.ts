// Libraries
import _ from 'lodash'

// Apis
import {
  writeLineProtocol,
  createTelegrafConfig,
  getTelegrafConfigs,
  updateTelegrafConfig,
  createScraperTarget,
  updateScraperTarget,
} from 'src/onboarding/apis/index'

// Utils
import {createNewPlugin} from 'src/onboarding/utils/pluginConfigs'

// Constants
import {
  pluginsByBundle,
  telegrafPluginsInfo,
} from 'src/onboarding/constants/pluginConfigs'

// Types
import {
  TelegrafPlugin,
  TelegrafPluginName,
  DataLoaderType,
  LineProtocolTab,
  Plugin,
  BundleName,
  ConfigurationState,
} from 'src/types/v2/dataLoaders'
import {AppState} from 'src/types/v2'
import {RemoteDataState} from 'src/types'
import {
  WritePrecision,
  TelegrafRequest,
  TelegrafPluginOutputInfluxDBV2,
} from 'src/api'
import {Dispatch} from 'redux'

type GetState = () => AppState

const DEFAULT_COLLECTION_INTERVAL = 10000

export type Action =
  | SetDataLoadersType
  | SetTelegrafConfigID
  | UpdateTelegrafPluginConfig
  | AddConfigValue
  | RemoveConfigValue
  | SetActiveTelegrafPlugin
  | SetLineProtocolBody
  | SetActiveLPTab
  | SetLPStatus
  | SetPrecision
  | UpdateTelegrafPlugin
  | AddPluginBundle
  | AddTelegrafPlugins
  | RemoveBundlePlugins
  | RemovePluginBundle
  | SetPluginConfiguration
  | SetConfigArrayValue
  | SetScraperTargetBucket
  | SetScraperTargetURL
  | SetScraperTargetID
  | ClearDataLoaders

interface SetDataLoadersType {
  type: 'SET_DATA_LOADERS_TYPE'
  payload: {type: DataLoaderType}
}

export const setDataLoadersType = (
  type: DataLoaderType
): SetDataLoadersType => ({
  type: 'SET_DATA_LOADERS_TYPE',
  payload: {type},
})

interface ClearDataLoaders {
  type: 'CLEAR_DATA_LOADERS'
}

export const clearDataLoaders = (): ClearDataLoaders => ({
  type: 'CLEAR_DATA_LOADERS',
})

interface UpdateTelegrafPluginConfig {
  type: 'UPDATE_TELEGRAF_PLUGIN_CONFIG'
  payload: {name: string; field: string; value: string}
}

export const updateTelegrafPluginConfig = (
  name: string,
  field: string,
  value: string
): UpdateTelegrafPluginConfig => ({
  type: 'UPDATE_TELEGRAF_PLUGIN_CONFIG',
  payload: {name, field, value},
})

interface UpdateTelegrafPlugin {
  type: 'UPDATE_TELEGRAF_PLUGIN'
  payload: {plugin: Plugin}
}

export const updateTelegrafPlugin = (plugin: Plugin): UpdateTelegrafPlugin => ({
  type: 'UPDATE_TELEGRAF_PLUGIN',
  payload: {plugin},
})

interface AddConfigValue {
  type: 'ADD_TELEGRAF_PLUGIN_CONFIG_VALUE'
  payload: {
    pluginName: string
    fieldName: string
    value: string
  }
}

export const addConfigValue = (
  pluginName: string,
  fieldName: string,
  value: string
): AddConfigValue => ({
  type: 'ADD_TELEGRAF_PLUGIN_CONFIG_VALUE',
  payload: {pluginName, fieldName, value},
})

interface RemoveConfigValue {
  type: 'REMOVE_TELEGRAF_PLUGIN_CONFIG_VALUE'
  payload: {
    pluginName: string
    fieldName: string
    value: string
  }
}

export const removeConfigValue = (
  pluginName: string,
  fieldName: string,
  value: string
): RemoveConfigValue => ({
  type: 'REMOVE_TELEGRAF_PLUGIN_CONFIG_VALUE',
  payload: {pluginName, fieldName, value},
})

interface SetConfigArrayValue {
  type: 'SET_TELEGRAF_PLUGIN_CONFIG_VALUE'
  payload: {
    pluginName: TelegrafPluginName
    field: string
    valueIndex: number
    value: string
  }
}

export const setConfigArrayValue = (
  pluginName: TelegrafPluginName,
  field: string,
  valueIndex: number,
  value: string
): SetConfigArrayValue => ({
  type: 'SET_TELEGRAF_PLUGIN_CONFIG_VALUE',
  payload: {pluginName, field, valueIndex, value},
})

interface SetTelegrafConfigID {
  type: 'SET_TELEGRAF_CONFIG_ID'
  payload: {id: string}
}

export const setTelegrafConfigID = (id: string): SetTelegrafConfigID => ({
  type: 'SET_TELEGRAF_CONFIG_ID',
  payload: {id},
})

interface AddPluginBundle {
  type: 'ADD_PLUGIN_BUNDLE'
  payload: {bundle: BundleName}
}

export const addPluginBundle = (bundle: BundleName): AddPluginBundle => ({
  type: 'ADD_PLUGIN_BUNDLE',
  payload: {bundle},
})

interface RemovePluginBundle {
  type: 'REMOVE_PLUGIN_BUNDLE'
  payload: {bundle: BundleName}
}

export const removePluginBundle = (bundle: BundleName): RemovePluginBundle => ({
  type: 'REMOVE_PLUGIN_BUNDLE',
  payload: {bundle},
})
interface AddTelegrafPlugins {
  type: 'ADD_TELEGRAF_PLUGINS'
  payload: {telegrafPlugins: TelegrafPlugin[]}
}

export const addTelegrafPlugins = (
  telegrafPlugins: TelegrafPlugin[]
): AddTelegrafPlugins => ({
  type: 'ADD_TELEGRAF_PLUGINS',
  payload: {telegrafPlugins},
})

interface RemoveBundlePlugins {
  type: 'REMOVE_BUNDLE_PLUGINS'
  payload: {bundle: BundleName}
}

export const removeBundlePlugins = (
  bundle: BundleName
): RemoveBundlePlugins => ({
  type: 'REMOVE_BUNDLE_PLUGINS',
  payload: {bundle},
})

interface SetScraperTargetBucket {
  type: 'SET_SCRAPER_TARGET_BUCKET'
  payload: {bucket: string}
}

export const setScraperTargetBucket = (
  bucket: string
): SetScraperTargetBucket => ({
  type: 'SET_SCRAPER_TARGET_BUCKET',
  payload: {bucket},
})

interface SetScraperTargetURL {
  type: 'SET_SCRAPER_TARGET_URL'
  payload: {url: string}
}

export const setScraperTargetURL = (url: string): SetScraperTargetURL => ({
  type: 'SET_SCRAPER_TARGET_URL',
  payload: {url},
})

interface SetScraperTargetID {
  type: 'SET_SCRAPER_TARGET_ID'
  payload: {id: string}
}

export const setScraperTargetID = (id: string): SetScraperTargetID => ({
  type: 'SET_SCRAPER_TARGET_ID',
  payload: {id},
})

export const addPluginBundleWithPlugins = (bundle: BundleName) => dispatch => {
  dispatch(addPluginBundle(bundle))
  const plugins = pluginsByBundle[bundle]
  dispatch(
    addTelegrafPlugins(
      plugins.map(p => {
        const isConfigured = !!telegrafPluginsInfo[p].fields
          ? ConfigurationState.Unconfigured
          : ConfigurationState.Configured

        return {
          name: p,
          active: false,
          configured: isConfigured,
        }
      })
    )
  )
}

export const removePluginBundleWithPlugins = (
  bundle: BundleName
) => dispatch => {
  dispatch(removePluginBundle(bundle))
  dispatch(removeBundlePlugins(bundle))
}

export const createOrUpdateTelegrafConfigAsync = (authToken: string) => async (
  dispatch,
  getState: GetState
) => {
  const {
    dataLoading: {
      dataLoaders: {telegrafPlugins},
      steps: {org, bucket, orgID},
    },
  } = getState()

  const telegrafConfigsFromServer = await getTelegrafConfigs(orgID)

  const influxDB2Out = {
    name: TelegrafPluginOutputInfluxDBV2.NameEnum.InfluxdbV2,
    type: TelegrafPluginOutputInfluxDBV2.TypeEnum.Output,
    config: {
      urls: ['http://127.0.0.1:9999'],
      token: authToken,
      organization: org,
      bucket,
    },
  }

  let plugins: Plugin[] = [influxDB2Out]
  telegrafPlugins.forEach(tp => {
    if (tp.configured === ConfigurationState.Configured) {
      plugins = [...plugins, tp.plugin || createNewPlugin(tp.name)]
    }
  })

  const telegrafRequest: TelegrafRequest = {
    name: 'new config',
    agent: {collectionInterval: DEFAULT_COLLECTION_INTERVAL},
    organizationID: orgID,
    plugins,
  }

  if (telegrafConfigsFromServer.length) {
    const id = _.get(telegrafConfigsFromServer, '0.id', '')

    await updateTelegrafConfig(id, telegrafRequest)
    dispatch(setTelegrafConfigID(id))
    return
  }

  const created = await createTelegrafConfig(telegrafRequest)
  dispatch(setTelegrafConfigID(created.id))
}

interface SetActiveTelegrafPlugin {
  type: 'SET_ACTIVE_TELEGRAF_PLUGIN'
  payload: {telegrafPlugin: string}
}

export const setActiveTelegrafPlugin = (
  telegrafPlugin: string
): SetActiveTelegrafPlugin => ({
  type: 'SET_ACTIVE_TELEGRAF_PLUGIN',
  payload: {telegrafPlugin},
})

interface SetPluginConfiguration {
  type: 'SET_PLUGIN_CONFIGURATION_STATE'
  payload: {telegrafPlugin: TelegrafPluginName}
}

export const setPluginConfiguration = (
  telegrafPlugin: TelegrafPluginName
): SetPluginConfiguration => ({
  type: 'SET_PLUGIN_CONFIGURATION_STATE',
  payload: {telegrafPlugin},
})

interface SetLineProtocolBody {
  type: 'SET_LINE_PROTOCOL_BODY'
  payload: {lineProtocolBody: string}
}

export const setLineProtocolBody = (
  lineProtocolBody: string
): SetLineProtocolBody => ({
  type: 'SET_LINE_PROTOCOL_BODY',
  payload: {lineProtocolBody},
})

interface SetActiveLPTab {
  type: 'SET_ACTIVE_LP_TAB'
  payload: {activeLPTab: LineProtocolTab}
}

export const setActiveLPTab = (
  activeLPTab: LineProtocolTab
): SetActiveLPTab => ({
  type: 'SET_ACTIVE_LP_TAB',
  payload: {activeLPTab},
})

interface SetLPStatus {
  type: 'SET_LP_STATUS'
  payload: {lpStatus: RemoteDataState}
}

export const setLPStatus = (lpStatus: RemoteDataState): SetLPStatus => ({
  type: 'SET_LP_STATUS',
  payload: {lpStatus},
})

interface SetPrecision {
  type: 'SET_PRECISION'
  payload: {precision: WritePrecision}
}

export const setPrecision = (precision: WritePrecision): SetPrecision => ({
  type: 'SET_PRECISION',
  payload: {precision},
})

export const writeLineProtocolAction = (
  org: string,
  bucket: string,
  body: string,
  precision: WritePrecision
) => async dispatch => {
  try {
    dispatch(setLPStatus(RemoteDataState.Loading))
    await writeLineProtocol(org, bucket, body, precision)
    dispatch(setLPStatus(RemoteDataState.Done))
  } catch (error) {
    dispatch(setLPStatus(RemoteDataState.Error))
  }
}

export const saveScraperTarget = () => async (
  dispatch: Dispatch<Action>,
  getState: GetState
) => {
  const {
    onboarding: {bucketID, orgID},
    dataLoading: {
      dataLoaders: {
        scraperTarget: {url, id},
      },
    },
  } = getState()

  try {
    if (id) {
      await updateScraperTarget(id, url, bucketID)
    } else {
      const newTarget = await createScraperTarget(url, orgID, bucketID)
      dispatch(setScraperTargetID(newTarget.id))
    }
  } catch (error) {
    console.error()
  }
}
