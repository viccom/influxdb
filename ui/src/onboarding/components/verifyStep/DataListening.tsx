// Libraries
import React, {PureComponent} from 'react'
import _ from 'lodash'

// Apis
import {executeQuery} from 'src/shared/apis/v2/query'

// Components
import {ErrorHandling} from 'src/shared/decorators/errors'
import {
  Button,
  ComponentColor,
  ComponentSize,
  ComponentStatus,
} from 'src/clockface'
import ConnectionInformation, {
  LoadingState,
} from 'src/onboarding/components/verifyStep/ConnectionInformation'

// Constants
import {StepStatus} from 'src/clockface/constants/wizard'

// Types
import {InfluxLanguage} from 'src/types/v2/dashboards'

export interface Props {
  bucket: string
  stepIndex: number
  onSetStepStatus: (index: number, status: StepStatus) => void
}

interface State {
  loading: LoadingState
  timePassedInSeconds: number
  secondsLeft: number
}

const MINUTE = 60000
const FETCH_WAIT = 5000
const SECONDS = 60
const TIMER_WAIT = 1000

@ErrorHandling
class DataListening extends PureComponent<Props, State> {
  private intervalID: NodeJS.Timer
  private startTime: number
  private timer: NodeJS.Timer

  constructor(props: Props) {
    super(props)

    this.state = {
      loading: LoadingState.NotStarted,
      timePassedInSeconds: 0,
      secondsLeft: SECONDS,
    }
  }

  public componentWillUnmount() {
    clearInterval(this.intervalID)
    clearInterval(this.timer)
    this.setState({
      timePassedInSeconds: 0,
      secondsLeft: SECONDS,
    })
  }

  public render() {
    return (
      <div className="wizard-step--body-streaming">
        {this.connectionInfo}
        {this.listenButton}
      </div>
    )
  }

  private get connectionInfo(): JSX.Element {
    const {loading} = this.state

    if (loading === LoadingState.NotStarted) {
      return
    }

    return (
      <ConnectionInformation
        loading={this.state.loading}
        bucket={this.props.bucket}
        countDownSeconds={this.state.secondsLeft}
      />
    )
  }

  private get listenButton(): JSX.Element {
    const {loading} = this.state

    if (loading === LoadingState.Loading || loading === LoadingState.Done) {
      return
    }

    return (
      <Button
        color={ComponentColor.Primary}
        text="Listen for Data"
        size={ComponentSize.Medium}
        onClick={this.handleClick}
        status={ComponentStatus.Default}
        titleText={'Listen for Data'}
      />
    )
  }

  private handleClick = (): void => {
    this.startTimer()
    this.setState({loading: LoadingState.Loading})
    this.startTime = Number(new Date())
    this.checkForData()
  }

  private checkForData = async (): Promise<void> => {
    const {bucket, onSetStepStatus, stepIndex} = this.props
    const {secondsLeft} = this.state
    const script = `from(bucket: "${bucket}")
      |> range(start: -1m)`

    let rowCount
    let timePassed

    try {
      const response = await executeQuery(
        '/api/v2/query',
        script,
        InfluxLanguage.Flux
      )
      rowCount = response.rowCount
      timePassed = Number(new Date()) - this.startTime
    } catch (err) {
      this.setState({loading: LoadingState.Error})
      onSetStepStatus(stepIndex, StepStatus.Incomplete)
      return
    }

    if (rowCount > 1) {
      this.setState({loading: LoadingState.Done})
      onSetStepStatus(stepIndex, StepStatus.Complete)
      return
    }

    if (timePassed >= MINUTE || secondsLeft <= 0) {
      this.setState({loading: LoadingState.NotFound})
      onSetStepStatus(stepIndex, StepStatus.Incomplete)
      return
    }
    this.intervalID = setTimeout(this.checkForData, FETCH_WAIT)
  }

  private startTimer() {
    this.setState({timePassedInSeconds: 0, secondsLeft: SECONDS})

    this.timer = setInterval(this.countDown, TIMER_WAIT)
  }

  private countDown = () => {
    const {secondsLeft} = this.state
    const secs = secondsLeft - 1
    this.setState({
      timePassedInSeconds: SECONDS - secs,
      secondsLeft: secs,
    })

    if (secs === 0) {
      clearInterval(this.timer)
    }
  }
}

export default DataListening
