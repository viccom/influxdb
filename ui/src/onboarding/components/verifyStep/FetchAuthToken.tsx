// Libraries
import React, {PureComponent} from 'react'
import _ from 'lodash'

// Components
import {Spinner} from 'src/clockface'
import {ErrorHandling} from 'src/shared/decorators/errors'

// Apis
import {getAuthorizationToken} from 'src/onboarding/apis/index'

// types
import {RemoteDataState} from 'src/types'

export interface Props {
  bucket: string
  username: string
  children: (authToken: string) => JSX.Element
}

interface State {
  loading: RemoteDataState
  authToken?: string
}

@ErrorHandling
class FetchAuthToken extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {loading: RemoteDataState.NotStarted}
  }

  public async componentDidMount() {
    const {username} = this.props

    this.setState({loading: RemoteDataState.Loading})
    const authToken = await getAuthorizationToken(username)

    this.setState({authToken, loading: RemoteDataState.Done})
  }

  public render() {
    return (
      <Spinner loading={this.state.loading}>
        {this.props.children(this.state.authToken)}
      </Spinner>
    )
  }
}

export default FetchAuthToken
