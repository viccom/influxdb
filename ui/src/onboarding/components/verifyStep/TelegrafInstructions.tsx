// Libraries
import React, {PureComponent} from 'react'
import _ from 'lodash'

// Decorator
import {ErrorHandling} from 'src/shared/decorators/errors'

// Components
import CopyText from 'src/shared/components/CopyText'

// Types
import {NotificationAction} from 'src/types'

export interface Props {
  notify: NotificationAction
  authToken: string
  configID: string
}

@ErrorHandling
class TelegrafInstructions extends PureComponent<Props> {
  public render() {
    const {notify, authToken, configID} = this.props
    const exportToken = `export INFLUX_TOKEN=${authToken || ''}`
    const configScript = `telegraf -config http://localhost:9999/api/v2/telegrafs/${configID ||
      ''}`
    return (
      <>
        <h3 className="wizard-step--title">Listen for Streaming Data</h3>
        <h5 className="wizard-step--sub-title">
          You have selected streaming data sources. Follow the instructions
          below to begin listening for incoming data.
        </h5>
        <div className="wizard-step--body">
          <h6>Install</h6>
          <p>
            You can download the binaries directly from the downloads page or
            from the releases section.{' '}
          </p>
          <h6>Start Data Stream</h6>
          <p>
            After installing the telegraf client, save this environment
            variable. run the following command.
          </p>
          <CopyText copyText={exportToken} notify={notify} />
          <p>Run the following command.</p>
          <CopyText copyText={configScript} notify={notify} />
        </div>
      </>
    )
  }
}

export default TelegrafInstructions
