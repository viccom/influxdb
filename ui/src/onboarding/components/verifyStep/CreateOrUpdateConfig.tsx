// Libraries
import React, {PureComponent} from 'react'
import _ from 'lodash'

// Components
import {Spinner} from 'src/clockface'
import {ErrorHandling} from 'src/shared/decorators/errors'

// Actions
import {createOrUpdateTelegrafConfigAsync} from 'src/onboarding/actions/dataLoaders'

// Constants
import {
  TelegrafConfigCreationSuccess,
  TelegrafConfigCreationError,
} from 'src/shared/copy/notifications'

// Types
import {RemoteDataState, NotificationAction} from 'src/types'

export interface Props {
  org: string
  authToken: string
  children: () => JSX.Element
  onSaveTelegrafConfig: typeof createOrUpdateTelegrafConfigAsync
  notify: NotificationAction
}

interface State {
  loading: RemoteDataState
}

@ErrorHandling
class CreateOrUpdateConfig extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {loading: RemoteDataState.NotStarted}
  }

  public async componentDidMount() {
    const {onSaveTelegrafConfig, authToken, notify} = this.props

    this.setState({loading: RemoteDataState.Loading})

    try {
      await onSaveTelegrafConfig(authToken)
      notify(TelegrafConfigCreationSuccess)

      this.setState({loading: RemoteDataState.Done})
    } catch (error) {
      notify(TelegrafConfigCreationError)
      this.setState({loading: RemoteDataState.Error})
    }
  }

  public render() {
    return (
      <Spinner loading={this.state.loading}>{this.props.children()}</Spinner>
    )
  }
}

export default CreateOrUpdateConfig
