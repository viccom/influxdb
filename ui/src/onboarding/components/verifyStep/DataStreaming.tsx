// Libraries
import React, {PureComponent} from 'react'
import _ from 'lodash'

// Components
import TelegrafInstructions from 'src/onboarding/components/verifyStep/TelegrafInstructions'
import CreateOrUpdateConfig from 'src/onboarding/components/verifyStep/CreateOrUpdateConfig'
import DataListening from 'src/onboarding/components/verifyStep/DataListening'

// Actions
import {createOrUpdateTelegrafConfigAsync} from 'src/onboarding/actions/dataLoaders'

// Constants
import {StepStatus} from 'src/clockface/constants/wizard'

// Decorator
import {ErrorHandling} from 'src/shared/decorators/errors'

// Types
import {NotificationAction} from 'src/types'

interface Props {
  notify: NotificationAction
  bucket: string
  org: string
  configID: string
  stepIndex: number
  authToken: string
  onSetStepStatus: (index: number, status: StepStatus) => void
  onSaveTelegrafConfig: typeof createOrUpdateTelegrafConfigAsync
}

@ErrorHandling
class DataStreaming extends PureComponent<Props> {
  public render() {
    const {
      authToken,
      org,
      configID,
      onSaveTelegrafConfig,
      onSetStepStatus,
      bucket,
      stepIndex,
      notify,
    } = this.props

    return (
      <>
        <CreateOrUpdateConfig
          org={org}
          notify={notify}
          authToken={authToken}
          onSaveTelegrafConfig={onSaveTelegrafConfig}
        >
          {() => (
            <TelegrafInstructions
              notify={notify}
              authToken={authToken}
              configID={configID}
            />
          )}
        </CreateOrUpdateConfig>

        <DataListening
          bucket={bucket}
          stepIndex={stepIndex}
          onSetStepStatus={onSetStepStatus}
        />
      </>
    )
  }
}

export default DataStreaming
