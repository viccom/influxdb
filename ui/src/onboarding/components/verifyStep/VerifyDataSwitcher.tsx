// Libraries
import React, {PureComponent} from 'react'

// Components
import {ErrorHandling} from 'src/shared/decorators/errors'
import DataStreaming from 'src/onboarding/components/verifyStep/DataStreaming'
import FetchAuthToken from 'src/onboarding/components/verifyStep/FetchAuthToken'

// Actions
import {createOrUpdateTelegrafConfigAsync} from 'src/onboarding/actions/dataLoaders'

// Constants
import {StepStatus} from 'src/clockface/constants/wizard'

// Types
import {DataLoaderType} from 'src/types/v2/dataLoaders'
import {NotificationAction, RemoteDataState} from 'src/types'
import StatusIndicator from 'src/onboarding/components/verifyStep/lineProtocol/StatusIndicator'

interface Props {
  notify: NotificationAction
  type: DataLoaderType
  org: string
  bucket: string
  username: string
  stepIndex: number
  telegrafConfigID: string
  onSaveTelegrafConfig: typeof createOrUpdateTelegrafConfigAsync
  onSetStepStatus: (index: number, status: StepStatus) => void
  onDecrementCurrentStep: () => void
  lpStatus: RemoteDataState
}

@ErrorHandling
export class VerifyDataSwitcher extends PureComponent<Props> {
  public render() {
    const {
      org,
      bucket,
      username,
      type,
      stepIndex,
      onSetStepStatus,
      telegrafConfigID,
      onSaveTelegrafConfig,
      notify,
      lpStatus,
    } = this.props

    switch (type) {
      case DataLoaderType.Streaming:
        return (
          <FetchAuthToken bucket={bucket} username={username}>
            {authToken => (
              <DataStreaming
                notify={notify}
                org={org}
                configID={telegrafConfigID}
                authToken={authToken}
                bucket={bucket}
                onSetStepStatus={onSetStepStatus}
                onSaveTelegrafConfig={onSaveTelegrafConfig}
                stepIndex={stepIndex}
              />
            )}
          </FetchAuthToken>
        )
      case DataLoaderType.LineProtocol:
        return (
          <StatusIndicator
            status={lpStatus}
            onClickRetry={this.handleClickRetry}
          />
        )
      default:
        return <div />
    }
  }

  private handleClickRetry = () => {
    this.props.onDecrementCurrentStep()
  }
}

export default VerifyDataSwitcher
