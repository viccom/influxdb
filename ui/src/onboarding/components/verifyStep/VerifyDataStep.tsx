// Libraries
import React, {PureComponent} from 'react'
import {connect} from 'react-redux'
import _ from 'lodash'

// Components
import {ErrorHandling} from 'src/shared/decorators/errors'
import VerifyDataSwitcher from 'src/onboarding/components/verifyStep/VerifyDataSwitcher'
import OnboardingButtons from 'src/onboarding/components/OnboardingButtons'
import FancyScrollbar from 'src/shared/components/fancy_scrollbar/FancyScrollbar'

// Actions
import {
  setActiveTelegrafPlugin,
  createOrUpdateTelegrafConfigAsync,
  setPluginConfiguration,
} from 'src/onboarding/actions/dataLoaders'

// Types
import {DataLoaderType, TelegrafPlugin} from 'src/types/v2/dataLoaders'
import {Form} from 'src/clockface'
import {NotificationAction, RemoteDataState} from 'src/types'
import {StepStatus} from 'src/clockface/constants/wizard'
import {AppState} from 'src/types/v2'
import {DataLoaderStepProps} from 'src/dataLoaders/components/DataLoadersWizard'

export interface OwnProps extends DataLoaderStepProps {
  notify: NotificationAction
  type: DataLoaderType
  telegrafConfigID: string
  telegrafPlugins: TelegrafPlugin[]
  onSetActiveTelegrafPlugin: typeof setActiveTelegrafPlugin
  onSetPluginConfiguration: typeof setPluginConfiguration
  onSaveTelegrafConfig: typeof createOrUpdateTelegrafConfigAsync
  stepIndex: number
  bucket: string
  username: string
  org: string
}

interface StateProps {
  lpStatus: RemoteDataState
}

export type Props = OwnProps & StateProps

@ErrorHandling
export class VerifyDataStep extends PureComponent<Props> {
  public componentDidMount() {
    const {type, onSetPluginConfiguration, telegrafPlugins} = this.props

    if (type === DataLoaderType.Streaming) {
      telegrafPlugins.forEach(tp => {
        onSetPluginConfiguration(tp.name)
      })
    }
  }

  public render() {
    const {
      bucket,
      username,
      telegrafConfigID,
      type,
      onSaveTelegrafConfig,
      onDecrementCurrentStepIndex,
      onSetStepStatus,
      stepIndex,
      notify,
      lpStatus,
      org,
    } = this.props

    return (
      <div className="onboarding-step wizard--skippable">
        <Form onSubmit={this.handleIncrementStep}>
          <div className="wizard-step--scroll-area">
            <FancyScrollbar autoHide={false}>
              <div className="wizard-step--scroll-content">
                <VerifyDataSwitcher
                  notify={notify}
                  type={type}
                  telegrafConfigID={telegrafConfigID}
                  onSaveTelegrafConfig={onSaveTelegrafConfig}
                  org={org}
                  bucket={bucket}
                  username={username}
                  onSetStepStatus={onSetStepStatus}
                  stepIndex={stepIndex}
                  onDecrementCurrentStep={onDecrementCurrentStepIndex}
                  lpStatus={lpStatus}
                />
              </div>
            </FancyScrollbar>
          </div>
          <OnboardingButtons
            onClickBack={this.handleDecrementStep}
            onClickSkip={this.jumpToCompletionStep}
            skipButtonText={'Skip'}
            showSkip={true}
            nextButtonText={'Finish'}
          />
        </Form>
      </div>
    )
  }

  private get previousStepName() {
    const {telegrafPlugins, type} = this.props

    if (type === DataLoaderType.Streaming) {
      return _.get(telegrafPlugins, `${telegrafPlugins.length - 1}.name`, '')
    }

    return type
  }

  private handleIncrementStep = () => {
    const {onSetStepStatus, type, lpStatus, onExit} = this.props
    const {currentStepIndex} = this.props

    if (
      type === DataLoaderType.LineProtocol &&
      lpStatus === RemoteDataState.Error
    ) {
      onSetStepStatus(currentStepIndex, StepStatus.Error)
    } else {
      onSetStepStatus(currentStepIndex, StepStatus.Complete)
    }

    onExit()
  }

  private handleDecrementStep = () => {
    const {
      telegrafPlugins,
      onSetActiveTelegrafPlugin,
      onDecrementCurrentStepIndex,
      onSetSubstepIndex,
      stepIndex,
      type,
    } = this.props

    if (type === DataLoaderType.Streaming) {
      onSetSubstepIndex(stepIndex - 1, telegrafPlugins.length - 1 || 0)
      onSetActiveTelegrafPlugin(this.previousStepName)
    } else {
      onDecrementCurrentStepIndex()
      onSetActiveTelegrafPlugin('')
    }
  }

  private jumpToCompletionStep = () => {
    const {onSetCurrentStepIndex, stepStatuses} = this.props

    onSetCurrentStepIndex(stepStatuses.length - 1)
  }
}

const mstp = ({
  dataLoading: {
    dataLoaders: {lpStatus},
  },
}: AppState): StateProps => ({
  lpStatus,
})

export default connect<StateProps, {}, OwnProps>(mstp)(VerifyDataStep)
