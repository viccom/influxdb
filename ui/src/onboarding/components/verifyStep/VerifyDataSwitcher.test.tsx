// Libraries
import React from 'react'
import {shallow} from 'enzyme'

// Components
import {VerifyDataSwitcher} from 'src/onboarding/components/verifyStep/VerifyDataSwitcher'

// Types
import {DataLoaderType} from 'src/types/v2/dataLoaders'
import {RemoteDataState} from 'src/types'

const setup = (override = {}) => {
  const props = {
    notify: jest.fn(),
    type: DataLoaderType.Empty,
    org: '',
    username: '',
    bucket: '',
    authToken: '',
    telegrafConfigID: '',
    onSaveTelegrafConfig: jest.fn(),
    stepIndex: 4,
    onSetStepStatus: jest.fn(),
    onDecrementCurrentStep: jest.fn(),
    lpStatus: RemoteDataState.NotStarted,
    ...override,
  }

  const wrapper = shallow(<VerifyDataSwitcher {...props} />)

  return {wrapper}
}

describe('Onboarding.Components.VerifyStep.VerifyDataSwitcher', () => {
  it('renders', () => {
    const {wrapper} = setup()

    expect(wrapper.exists()).toBe(true)
  })

  describe('If data type is streaming', () => {
    it('renders the DataStreaming component', () => {
      const {wrapper} = setup({type: DataLoaderType.Streaming})

      expect(wrapper).toMatchSnapshot()
    })
  })
})
