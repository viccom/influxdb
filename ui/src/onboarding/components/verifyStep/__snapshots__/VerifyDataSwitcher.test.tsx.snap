// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Onboarding.Components.VerifyStep.VerifyDataSwitcher If data type is streaming renders the DataStreaming component 1`] = `
<FetchAuthToken
  bucket=""
  username=""
>
  <Component />
</FetchAuthToken>
`;
